### DISCUSSION TITLE
Some questions about nn.conv.han_conv

### DISCUSSION THREAD

**Discussion by ZRH0308** [pyg-team/pytorch_geometric#6364] [Q&A] [answered]
Hi, 

I'm a bit confused about the function **group**   in   **nn.conv.han_conv**

```
def group(
    xs: List[Tensor],
    q: nn.Parameter,
    k_lin: nn.Module,
) -> Tuple[OptTensor, OptTensor]:

    if len(xs) == 0:
        return None, None
    else:
        num_edge_types = len(xs)
        out = torch.stack(xs)
        if out.numel() == 0:
            return out.view(0, out.size(-1)), None
        attn_score = (q * torch.tanh(k_lin(out)).mean(1)).sum(-1)
        attn = F.softmax(attn_score, dim=0)
        out = torch.sum(attn.view(num_edge_types, 1, -1) * out, dim=0)
        return out, attn
```

**Quetion1**:  When we calling the function **group**, Since the `xs` is `outs` of only one node type, why we could use `F.softmax(attn_score, dim=0)` to obtain the attention weight for this node type here ? If I understand correctly, here we also need attn_score from other node types,  just like the eq.8 in the original paper. [https://arxiv.org/abs/1903.07293v1](url)  

**Quetion2**:  Could you please explain to me the meaning of `.mean(1)` after `torch.tanh(k_lin(out))` ? In the original paper, it seems that attn_score for one node type are just obtained by averaging the result of dot q & MLP(outs).

These questions really bother me. Is there something wrong with my understanding? Could you please explain to me? Thank you very much !

**Comment by rusty1s** [comment, marked as answer]
1: `xs` contains the embeddings per edge type or in the case of `HANConv`, per metapath. 

2: We are averaging the scores over all nodes which refers to Eq. 7.

Hope this clarifies your doubts.

**Reply by ZRH0308** [reply]
Oh, I see! I had mistakenly thought that `node_type` refers to metapath.

Thank you very much for clarifying all my doubts accurately.
